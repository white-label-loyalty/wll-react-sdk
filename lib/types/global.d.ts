// Global type definitions for React Native environment

declare const __DEV__: boolean;
