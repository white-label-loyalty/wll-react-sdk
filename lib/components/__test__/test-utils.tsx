import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
