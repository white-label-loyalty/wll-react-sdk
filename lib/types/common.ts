import { ImageProps, ViewStyle } from 'react-native';

export type ImagePropsNoSource = Omit<ImageProps, 'source'>;

export type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type Justify =
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'evenly';

export type Align = 'start' | 'end' | 'center' | 'stretch' | 'baseline';

export type LayoutProps = {
  children: React.ReactNode;
  justify?: Justify;
  align?: Align;
  direction?: FlexDirection;
  style?: ViewStyle;
};

export const justifyMap: Record<Justify, ViewStyle['justifyContent']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  between: 'space-between',
  around: 'space-around',
  evenly: 'space-evenly',
} as const;

export const alignMap: Record<Align, ViewStyle['alignItems']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;
