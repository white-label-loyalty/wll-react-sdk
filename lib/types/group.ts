import { TSection } from './section';

export type TGroup = {
  name: string;
  active: boolean;
  id: string;
  createdAt: string;
  updatedAt: string;
  sections: TSection[];
};
