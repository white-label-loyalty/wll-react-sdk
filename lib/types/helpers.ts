/**
 * Makes a configuration type with required and optional fields,
 * providing default values for optional fields.
 */
export type SafeConfig<T, R extends keyof T> = Required<Pick<T, R>> &
  Partial<Omit<T, R>>;

/**
 * Adds optional children to any component props.
 */
export type WithChildren<T = {}> = T & { children?: React.ReactNode };
