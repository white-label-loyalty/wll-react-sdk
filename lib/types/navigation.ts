export type NavigationType = 'external' | 'internal' | 'modal';

type NavigationHandlerParams = {
  target: string;
  windowTarget: string;
};

export type NavigationConfig = {
  navigationHandlers?: {
    external?: (params: NavigationHandlerParams) => Promise<void>;
    internal?: (params: NavigationHandlerParams) => Promise<void>;
    modal?: (params: NavigationHandlerParams) => Promise<void>;
  };
  external?: (params: NavigationHandlerParams) => Promise<void>;
  internal?: (params: NavigationHandlerParams) => Promise<void>;
  modal?: (params: NavigationHandlerParams) => Promise<void>;
  baseUrl?: string;
};
