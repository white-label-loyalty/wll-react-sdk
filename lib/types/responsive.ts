export type DimensionMode = 'desktop' | 'tablet' | 'mobile';

export type WindowDimensions = {
  width: number;
  height: number;
};

export type ResponsiveContextType = {
  dimensionMode: DimensionMode;
  isTablet: boolean;
  isMobile: boolean;
  isDesktop: boolean;
};
