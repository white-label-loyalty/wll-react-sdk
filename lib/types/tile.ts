export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
