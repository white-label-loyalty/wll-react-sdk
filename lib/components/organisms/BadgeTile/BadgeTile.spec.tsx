import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import React from 'react';
import { BadgeTileType, TileHeight, TileType } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { BadgeTile } from './index';

const BadgeTileMock = {
  id: 'latest-badge-none',
  type: TileType.Badge,
  active: true,
  createdAt: '2024-08-06T08:53:24.307Z',
  updatedAt: '2024-08-06T08:53:24.307Z',
  tileHeight: TileHeight.Full,
  priority: 0,
  configuration: {
    type: BadgeTileType.Latest,
    badgeId: '900a2477-95c4-4c42-ae2d-3795e7f0f5f2',
    internalName: 'Top Spender',
    name: 'Top Spender',
    description: 'Spent Â£100 on 5 Separate transactions',
    artworkUrl: 'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
    priority: 0,
    status: 'ACTIVE',
    createdAt: '2024-08-06T08:53:24.307Z',
    updatedAt: '2024-08-06T08:53:24.307Z',
    awardedDatePrefix: 'Awarded',
    emptyBadgeMessage: "You haven't earned any badges yet.",
    badgeNotEarnedMessage: 'Badge not earned yet',
    emptyBadgeArtworkUrl:
      'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
    count: 0,
  },
};

describe('<BadgeTile />', () => {
  it('matches snapshot', () => {
    const { container } = render(<BadgeTile tile={BadgeTileMock} />);
    expect(container).toMatchSnapshot();
  });

  describe('Rendering States', () => {
    it('returns null when tile is not active', () => {
      const inactiveTile = {
        ...BadgeTileMock,
        active: false,
      };
      const { container } = render(<BadgeTile tile={inactiveTile} />);
      expect(container).toBeEmptyDOMElement();
    });

    it('returns null when tile height is not full', () => {
      const halfHeightTile = {
        ...BadgeTileMock,
        tileHeight: TileHeight.Half,
      };
      const { container } = render(<BadgeTile tile={halfHeightTile} />);
      expect(container).toBeEmptyDOMElement();
    });

    it('renders correctly for Latest badge type with no badges earned', () => {
      render(<BadgeTile tile={BadgeTileMock} />);
      expect(
        screen.getByText(BadgeTileMock.configuration.emptyBadgeMessage)
      ).toBeInTheDocument();
      expect(screen.getByTestId('badge-tile-media')).toBeInTheDocument();
    });

    it('renders correctly for Latest badge type with badges earned', () => {
      const earnedBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          count: 1,
        },
      };
      render(<BadgeTile tile={earnedBadge} />);
      expect(
        screen.getByText(earnedBadge.configuration.name)
      ).toBeInTheDocument();
      expect(
        screen.getByText(earnedBadge.configuration.description)
      ).toBeInTheDocument();
      expect(screen.getByTestId('badge-tile-media')).toBeInTheDocument();
    });
  });

  describe('Badge Types', () => {
    it('renders specific badge type correctly when not earned', () => {
      const specificBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          type: BadgeTileType.Specific,
          count: 0,
        },
      };
      render(<BadgeTile tile={specificBadge} />);
      expect(
        screen.getByText(specificBadge.configuration.badgeNotEarnedMessage)
      ).toBeInTheDocument();
    });

    it('renders specific badge type correctly when earned once', () => {
      const specificBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          type: BadgeTileType.Specific,
          count: 1,
        },
      };
      render(<BadgeTile tile={specificBadge} />);
      expect(
        screen.getByText(specificBadge.configuration.name)
      ).toBeInTheDocument();
      expect(
        screen.getByText(specificBadge.configuration.description)
      ).toBeInTheDocument();
    });

    it('renders Lock icon when not earned', () => {
      const specificBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          type: BadgeTileType.Specific,
          count: 0,
        },
      };
      render(<BadgeTile tile={specificBadge} />);
      expect(
        screen.getByText(specificBadge.configuration.badgeNotEarnedMessage)
      ).toBeInTheDocument();
      expect(
        screen.getByTestId('badge-tile-status-locked')
      ).toBeInTheDocument();
    });

    it('renders specific badge type correctly with multiple earned', () => {
      const specificBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          type: BadgeTileType.Specific,
          count: 3,
        },
      };
      render(<BadgeTile tile={specificBadge} />);
      expect(
        screen.getByText(specificBadge.configuration.name)
      ).toBeInTheDocument();
      expect(screen.getByText('3x')).toBeInTheDocument();
    });
  });

  describe('Badge Media', () => {
    it('renders badge artwork when provided', () => {
      render(<BadgeTile tile={BadgeTileMock} />);
      expect(screen.getByTestId('badge-tile-media')).toBeInTheDocument();
    });

    it('renders empty badge artwork when no badges earned', () => {
      const emptyBadge = {
        ...BadgeTileMock,
        configuration: {
          ...BadgeTileMock.configuration,
          count: 0,
        },
      };
      render(<BadgeTile tile={emptyBadge} />);
      const media = screen.getByTestId('badge-tile-media');
      expect(media).toBeInTheDocument();

      const progressiveImage = media.querySelector('[src]');
      expect(progressiveImage).toHaveAttribute(
        'src',
        emptyBadge.configuration.emptyBadgeArtworkUrl
      );
    });
  });
});
