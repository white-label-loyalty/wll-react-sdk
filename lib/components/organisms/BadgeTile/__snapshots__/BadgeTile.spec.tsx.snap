// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadgeTile /> matches snapshot 1`] = `
<div>
  <button
    aria-disabled="true"
    aria-label="Tile"
    class="css-view-175oi2r r-pointerEvents-12vffkv r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
    disabled=""
    role="button"
    style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
    tabindex="-1"
    type="button"
  >
    <div
      class="css-view-175oi2r r-alignItems-1awozwy r-flexBasis-1t2qqvi r-justifyContent-1777fci r-marginBottom-1ifxtd0 r-position-bnwqim"
      data-testid="badge-tile-media"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-justifyContent-1777fci r-position-bnwqim r-height-1pi2tsx r-width-13qz1uu"
        style="background-color: rgba(57, 46, 215, 0.2);"
      >
        <div
          class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="background-color: rgba(57, 46, 215, 0.2); opacity: 1;"
        />
        <div
          aria-label="Badge Top Spender (not earned)"
          class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="opacity: 0;"
        >
          <div
            class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-ehq7j7"
            style="background-image: url(https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/);"
          />
          <img
            alt="Badge Top Spender (not earned)"
            class="css-accessibilityImage-9pa8cd"
            draggable="false"
            src="https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/"
          />
        </div>
      </div>
    </div>
    <div
      class="css-view-175oi2r r-alignItems-1habvwh r-flex-13awgt0 r-flexDirection-eqz5dr r-justifyContent-1wtj0ep r-paddingBottom-kzbkwu r-paddingInline-3o4zer"
    >
      <div
        aria-label="Badge title: You haven't earned any badges yet."
        class="css-view-175oi2r"
        data-testid="badge-tile-title"
      >
        <div
          class="css-text-146c3p1 r-WebkitBoxOrient-8akbws r-display-krxsd3 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-marginBottom-5oul0u"
          dir="auto"
          style="font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif; color: rgb(0, 0, 0); font-size: 24px; font-weight: bold;"
        >
          You haven't earned any badges yet.
        </div>
      </div>
      <div
        class="css-view-175oi2r"
        style="flex: 1;"
      />
    </div>
  </button>
</div>
`;
